### PR TITLE
Refactor how tasks are cleaned up

### PR DESCRIPTION
**Pull request by pipermerriam**
## What was wrong?

The way that tasks were being cleaned up left room for edge cases where some cleanup might fail.

## How was it fixed?

Created a fancy async context manager to do it in a more reliable and granular way that ensures each individual task is isolated during cleanup and an exception during cleanup of one task won't cause the cleanup of the others to fail.

#### Cute Animal Picture

2BE9013F00000578-0-image-a-5_1441209978133

### CODE REVIEW

File: asyncio_run_in_process/_utils.py [hunk 1]
```diff
@@ -87,3 +94,33 @@ def __reduce__(self) -> Any:
 def rebuild_exc(exc, tb):  # type: ignore
     exc.__cause__ = RemoteTraceback(tb)
     return exc
+
+
+def cleanup_tasks(task: 'asyncio.Future[Any]',
+                  *others: 'asyncio.Future[Any]',
```
Comment (gsalgado) [inline asyncio_run_in_process/_utils.py]: It'd be nice to have a docstring explaining how the different arguments are treated. I mean, it looks like they aren't treated the same, but I'm not really sure what's the difference

File: asyncio_run_in_process/run_in_process.py [hunk 2]
```diff
@@ -240,51 +241,34 @@ def _update_subprocess_kwargs(subprocess_kwargs: Optional['SubprocessKwargs'],
...
+            # If a keyboard interrupt is encountered relay it to the
+            # child process and then give it a moment to cleanup before
+            # re-raising
             try:
-                await asyncio.wait_for(proc.wait(), timeout=2)
-            except asyncio.TimeoutError:
-                pass
+                proc.send_signal(signal.SIGINT)
+                try:
+                    await asyncio.wait_for(proc.wait(), timeout=2)
```
Comment (gsalgado) [inline asyncio_run_in_process/run_in_process.py:265]: Should we at least use a constant for this timeout?

**Review by gsalgado — APPROVED** [review]
It looks ok to me, although it pokes deep in the guts of a nontrivial piece of code and I worry I don't understand all the implications here. Also, no tests had to be modified, which makes me think most of this is untested code :-(

File: asyncio_run_in_process/run_in_process.py (hunk 2 above)
Comment (pipermerriam) [reply asyncio_run_in_process/run_in_process.py:265]: oops, these changes weren't supposed to be included in this PR, they were intended for #6

File: asyncio_run_in_process/_utils.py (hunk 1 above)
Comment (pipermerriam) [reply asyncio_run_in_process/_utils.py]: Adding a docstring.

As for your confusion.  This *could* be written as `def cleanup_tasks(*tasks)`.  However, by doing it with the extra initial positional argument, it:

1. ensures that there is at least 1 task without having to validate `len(tasks) >= 1`
2. eliminates the need to pull off the first task via `tasks[0]`.

**Pull request merged**